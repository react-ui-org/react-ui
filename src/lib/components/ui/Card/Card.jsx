import PropTypes from 'prop-types';
import React from 'react';
import getRootColorClassName from '../../../helpers/getRootColorClassName';
import { withProviderContext } from '../../../provider';
import styles from './Card.scss';

export const Card = ({
  children,
  dense,
  disabled,
  id,
  raised,
  color,
}) => (
  <div
    className={[
      styles.root,
      getRootColorClassName(color, styles),
      dense ? styles.rootDense : '',
      raised ? styles.rootRaised : '',
      disabled ? styles.isDisabled : '',
    ].join(' ')}
    id={id}
  >
    {children}
  </div>
);

Card.defaultProps = {
  color: 'light',
  dense: false,
  disabled: false,
  id: undefined,
  raised: false,
};

Card.propTypes = {
  /**
   * Slot for individual card elements that build up the inner layout:
   * * `CardBody`
   * * `CardFooter`
   * * `ScrollView`
   */
  children: PropTypes.oneOfType([
    PropTypes.element,
    PropTypes.arrayOf(PropTypes.element),
  ]).isRequired,
  /**
   * [Color variant](/foundation/colors#component-colors) to clarify importance and meaning of the card.
   */
  color: PropTypes.oneOf(
    ['primary', 'secondary', 'success', 'warning', 'danger', 'help', 'info', 'note', 'light', 'dark'],
  ),
  /**
   * Make the card more compact.
   */
  dense: PropTypes.bool,
  /**
   * If `true`, the card will be disabled.
   */
  disabled: PropTypes.bool,
  /**
   * ID of the root HTML element.
   */
  id: PropTypes.string,
  /**
   * Add shadow to pull the card above surface.
   */
  raised: PropTypes.bool,
};

export const CardWithContext = withProviderContext(Card, 'Card');

export default CardWithContext;
