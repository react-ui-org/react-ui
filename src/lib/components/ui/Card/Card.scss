@import 'mixins';

.root {
  @include card();
}

.flat {
  @include card-type(flat);
}

.bordered {
  @include card-type(bordered);
}

.warning {
  @include card-type(warning);
}

.error {
  @include card-type(error);
}

.isRootDisabled {
  @include card-disabled();
}

.isRootInList {
  @include card-in-list();
}

.isRootRaised {
  @include card-raised();
}
