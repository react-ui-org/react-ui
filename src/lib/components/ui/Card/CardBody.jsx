import PropTypes from 'prop-types';
import React from 'react';
import styles from './Card.scss';

const CardBody = (props) => {
  const {
    children,
  } = props;

  return (
    <div className={styles.body}>
      {children}
    </div>
  );
};

CardBody.propTypes = {
  children: PropTypes.node.isRequired,
};

export default CardBody;
