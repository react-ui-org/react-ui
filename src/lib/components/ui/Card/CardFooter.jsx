import PropTypes from 'prop-types';
import React from 'react';
import styles from './Card.scss';

const CardFooter = (props) => {
  const {
    children,
  } = props;

  return (
    <div className={styles.footer}>
      {children}
    </div>
  );
};

CardFooter.propTypes = {
  children: PropTypes.node.isRequired,
};

export default CardFooter;
