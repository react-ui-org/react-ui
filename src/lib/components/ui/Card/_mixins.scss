@import 'theme';

@mixin card-type($type) {
  $type-properties: map-get($card-types, $type);

  border-color: map-get($type-properties, border-color);
}
