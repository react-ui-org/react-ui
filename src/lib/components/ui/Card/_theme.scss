$card-padding: var(--rui-card-padding);
$card-dense-padding: var(--rui-card-dense-padding);
$card-background-color: var(--rui-card-background-color);
$card-border-width: var(--rui-card-border-width);
$card-border-radius: var(--rui-card-border-radius);

$card-types: (
  success: (
    border-color: var(--rui-card-success-border-color),
  ),
  warning: (
    border-color: var(--rui-card-warning-border-color),
  ),
  error: (
    border-color: var(--rui-card-error-border-color),
  ),
  help: (
    border-color: var(--rui-card-help-border-color),
  ),
  info: (
    border-color: var(--rui-card-info-border-color),
  ),
  note: (
    border-color: var(--rui-card-note-border-color),
  ),
);

$card-variants: (
  flat: (
    border-color: var(--rui-card-flat-border-color),
  ),
  bordered: (
    border-color: var(--rui-card-bordered-border-color),
  ),
);

$card-disabled: (
  background-color: var(--rui-card-disabled-background-color),
  opacity: var(--rui-card-disabled-opacity),
);

$card-raised: (
  box-shadow: var(--rui-card-raised-box-shadow),
);
