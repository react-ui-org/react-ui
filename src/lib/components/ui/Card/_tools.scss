@use 'theme';

@mixin card-type($type) {
  $type-properties: map-get(theme.$types, $type);

  border-color: map-get($type-properties, border-color);
}

@mixin card-variant($variant) {
  $variant-properties: map-get(theme.$variants, $variant);

  border-color: map-get($variant-properties, border-color);
}
