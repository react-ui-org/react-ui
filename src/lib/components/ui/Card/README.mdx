---
name: Card
menu: 'Components: UI'
route: /components/ui/card
---

# Card

import {
  Playground,
  Props,
} from 'docz'
import { Button } from '../Button/Button'
import { Card } from './Card'
import CardBody from './CardBody'
import CardFooter from './CardFooter'

Cards contain content and actions about a single subject.

## Basic Usage

To implement the Card component, you need to import at least Card and
[CardBody](#cardbody) components:

```js
import { Card, CardBody } from '@react-ui-org/react-ui';
```

And use it:

<Playground>
  <Card>
    <CardBody>
      Hello!
    </CardBody>
  </Card>
</Playground>

See [API](#api) for all available options.

## General Guidelines

- The default (flat) card is **designed for a non-white background.** You may
  want to either use a different [variant](#variants) or
  [customize](/customize/theming) the default appearance to make it stand
  out on white surfaces.

- The [CardBody](#cardbody) component is the only required element in a card.
  [CardFooter](#cardfooter) is optional.

- Use **medium or low-emphasis buttons** when there are more cards, eg. in a
  grid. It will help you keep the UI clean and easy to scan.

## Card Composition

Aside from the mandatory [CardBody](#cardbody) element, you can add a
[CardFooter](#cardfooter) to better separate your card's actions from the rest
of the content.

Import all necessary components:

```js
import { Card, CardBody, CardFooter } from '@react-ui-org/react-ui';
````

And use them:

<Playground>
  <Card>
    <CardBody>
      Hello! I&apos;m card with a footer.
    </CardBody>
    <CardFooter>
      <Button label="Read more" priority="outline" />
    </CardFooter>
  </Card>
</Playground>

ðŸ‘‰ CardFooter is required in case you need to align the actions of multiple
cards with varying height of content.

## Raised Card

Add optional shadow to lift the card above surface.

<Playground>
  <Card raised>
    <CardBody>
      Hello! I&apos;m a raised card with a footer.
    </CardBody>
    <CardFooter>
      <Button label="Read more" priority="outline" />
    </CardFooter>
  </Card>
</Playground>

## Bordered Card

Use a border when you need to visually separate the card from its background.

<Playground>
  <Card type="bordered">
    <CardBody>
      Hello! I&apos;m card with a border and a footer.
    </CardBody>
    <CardFooter>
      <Button label="Read more" priority="outline" />
    </CardFooter>
  </Card>
  <Card type="bordered" raised>
    <CardBody>
      Hello! I&apos;m a raised card with a border and a footer.
    </CardBody>
    <CardFooter>
      <Button label="Read more" priority="outline" />
    </CardFooter>
  </Card>
</Playground>

## Dense Card

Use a dense card when you need to save space on screen. Other elements in the
card should be also smaller to keep the card contained and easy to scan.

<Playground>
  <Card dense raised>
    <CardBody>
      <small>Hello! I&apos;m a dense card. Everything is smaller here.</small>
    </CardBody>
    <CardFooter>
      <Button label="Read more" priority="outline" size="small" />
    </CardFooter>
  </Card>
</Playground>

## Variants

All [UI colors](/foundation/colors#ui-colors) (success, warning, error, help,
info, and note) are supported by cards to cover various needs of your app. The
bordered card type is forced automatically to make sure the border is visible.

<Playground>
  <Card type="success">
    <CardBody>
      Hello! I&apos;m success variant of card.
    </CardBody>
    <CardFooter>
      <Button label="Read more" priority="outline" variant="success" />
    </CardFooter>
  </Card>
  <Card type="warning">
    <CardBody>
      Hello! I&apos;m warning variant of card.
    </CardBody>
    <CardFooter>
      <Button label="Read more" priority="outline" variant="warning" />
    </CardFooter>
  </Card>
  <Card type="error">
    <CardBody>
      Hello! I&apos;m error variant of card.
    </CardBody>
    <CardFooter>
      <Button label="Read more" priority="outline" variant="danger" />
    </CardFooter>
  </Card>
  <Card type="help">
    <CardBody>
      Hello! I&apos;m help variant of card.
    </CardBody>
    <CardFooter>
      <Button label="Read more" priority="outline" variant="dark" />
    </CardFooter>
  </Card>
  <Card type="info">
    <CardBody>
      Hello! I&apos;m info variant of card.
    </CardBody>
    <CardFooter>
      <Button label="Read more" priority="outline" variant="dark" />
    </CardFooter>
  </Card>
  <Card type="note">
    <CardBody>
      Hello! I&apos;m note variant of card.
    </CardBody>
    <CardFooter>
      <Button label="Read more" priority="outline" variant="dark" />
    </CardFooter>
  </Card>
</Playground>

## States

### Disabled State

An entire card can appear disabled. However, you'll still need to manually
disable its interactive elements to disallow user's interaction.

<Playground>
  <Card disabled>
    <CardBody>
      Hello! I&apos;m a disabled card.
    </CardBody>
    <CardFooter>
      <Button label="Read more" priority="outline" disabled />
    </CardFooter>
  </Card>
  <Card disabled raised>
    <CardBody>
      Hello! I&apos;m a disabled raised card.
    </CardBody>
    <CardFooter>
      <Button label="Read more" priority="outline" disabled />
    </CardFooter>
  </Card>
  <Card variant="bordered" disabled>
    <CardBody>
      Hello! I&apos;m a disabled bordered card.
    </CardBody>
    <CardFooter>
      <Button label="Read more" priority="outline" disabled />
    </CardFooter>
  </Card>
  <Card type="success" disabled>
    <CardBody>
      Hello! I&apos;m a disabled success variant of card.
    </CardBody>
    <CardFooter>
      <Button label="Read more" priority="outline" variant="success" disabled />
    </CardFooter>
  </Card>
</Playground>

## API

<Props of={Card} />

## CardBody

CardBody is the only required element in a card. See
[Card Composition](#card-composition) for all details.

<Props of={CardBody} />

## CardFooter

Separate your card actions with CardFooter. See
[Card Composition](#card-composition) for all details.

<Props of={CardFooter} />

---

Next: [CheckboxField â†’](/components/ui/checkbox-field)
