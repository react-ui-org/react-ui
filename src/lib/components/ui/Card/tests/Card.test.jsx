import React from 'react';
import { shallow } from 'enzyme';
import { Card } from '../Card';

describe('rendering', () => {
  it('renders correctly', () => {
    const tree = shallow(<Card><p>Card content</p></Card>);

    expect(tree).toMatchSnapshot();
  });

  it('renders correctly error', () => {
    const tree = shallow(<Card type="error"><p>Card content</p></Card>);

    expect(tree).toMatchSnapshot();
  });

  it('renders correctly disabled', () => {
    const tree = shallow(<Card disabled><p>Card content</p></Card>);

    expect(tree).toMatchSnapshot();
  });

  it('renders correctly raised', () => {
    const tree = shallow(<Card raised><p>Card content</p></Card>);

    expect(tree).toMatchSnapshot();
  });

  it('renders correctly with all props', () => {
    const tree = shallow(
      <Card
        dense
        disabled
        id="custom-id"
        raised
        type="warning"
      >
        <p>Card content</p>
      </Card>,
    );

    expect(tree).toMatchSnapshot();
  });
});
