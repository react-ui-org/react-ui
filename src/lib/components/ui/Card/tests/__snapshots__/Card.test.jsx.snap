// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rendering renders correctly 1`] = `
<div
  className="root
        isTypeFlat"
>
  <p>
    Card content
  </p>
</div>
`;

exports[`rendering renders correctly disabled 1`] = `
<div
  className="root
        isTypeFlat
        
        isDisabled"
>
  <p>
    Card content
  </p>
</div>
`;

exports[`rendering renders correctly error 1`] = `
<div
  className="root
        isTypeError"
>
  <p>
    Card content
  </p>
</div>
`;

exports[`rendering renders correctly raised 1`] = `
<div
  className="root
        isTypeFlat
        
        
        
        isRaised"
>
  <p>
    Card content
  </p>
</div>
`;

exports[`rendering renders correctly with all props 1`] = `
<div
  className="root
        isTypeWarning
        isDense
        isDisabled
        
        isRaised"
  id="custom-id"
>
  <p>
    Card content
  </p>
</div>
`;
